Custom health checks

benhoyt opened the pull request:
Not the real thing, this is a PR on the `benhoyt` fork of the custom health checks work against the auto-restart branch. The real PR will be at https://github.com/canonical/pebble/pull/84

benhoyt committed:
Rename "use-terminal" to "terminal"; add separate "interactive" parameter (#81)

Also refactor WebsocketExecMirror so it only does the one direction,
and the caller can call WebsocketRecvStream separately for the other
direction. This was poorly factored to begin with.

benhoyt committed:
Check config, manager, and checkers (not wired up yet)

benhoyt committed:
First pass at checks validation and wiring up checks (not final!)

benhoyt committed:
Add /v1/checks API; add level and name filtering

benhoyt committed:
Add checks Go client and "pebble checks" CLI

benhoyt committed:
Wire up check manager to handle plan updates

benhoyt committed:
Partially wire on on-check-failure action (still need plan fields)

benhoyt committed:
Add on-check-failure to plan YAML

benhoyt committed:
Merge branch 'auto-restart' into custom-health-checks

Conflicts were:
- internal/overlord/servstate/handlers.go
- internal/overlord/servstate/manager.go
- internal/plan/plan.go

benhoyt committed:
Move overall check validation to combined level

benhoyt committed:
Add tests for the checkers: http, tcp, and exec

Also add a bit more validation to checks plan config. And ensure we
call shlex.Split to validate command up front, as the code says we do!

benhoyt committed:
Test the CheckManager; test of newChecker

benhoyt committed:
Change default number of check failures to 3 to avoid possible flapping

hpidcock commented on internal/overlord/checkstate/checkers.go:
```diff
@@ -0,0 +1,165 @@
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		if ctx.Err() == context.DeadlineExceeded {
+			err = fmt.Errorf("exec check timed out")
```
if the check timed out it should be SIGKILLed before returning from this method.

Or at least: SIGTERM, start 30second timer, SIGKILL

benhoyt replied:
I don't think there's any need for this: per `exec.CommandContext` [docs](https://pkg.go.dev/os/exec#CommandContext), the exec library already kills the process (SIGKILL) "if the context becomes done before the command completes on its own" (that would apply to cancellation or timeout).

hpidcock commented on internal/overlord/checkstate/checkers.go:
```diff
@@ -0,0 +1,165 @@
+		}
+		if len(output) > 0 {
+			const maxLength = 1024
+			if len(output) > maxLength {
```
I think this needs a unit test

benhoyt replied:
Good call -- added.

hpidcock commented on internal/overlord/checkstate/manager.go:
```diff
@@ -0,0 +1,216 @@
+		if level != plan.UnsetLevel && level != check.config.Level {
+			continue
+		}
+		if len(names) > 0 && !containsString(names, check.config.Name) {
```
looks like containsString is used once, probably just inline it here

benhoyt replied:
I think it's a bit nicer as a stand-alone function, as it allows a simpler early return rather than something like `break containsLoop`. However, I found that we already have a `strutil.ListContains` function, so I changed it to use that. Likely in Go 1.19 we'll be able to write `slices.Contains` (in Go 1.18 `slices` will be an experimental generics-based package).

hpidcock commented on internal/overlord/checkstate/manager.go:
```diff
@@ -0,0 +1,216 @@
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.Timeout.Value)
+	defer cancel()
+	c.mutex.Lock()
+	c.cancel = cancel
```
checkData should have its own context.Context

`c.ctx, c.cancel = context.WithCancel(context.Background())`

then check()'s ctx is single purpose
```
cox, cancel := context.WithTimeout(c.ctx, c.config.Timeout.Value)
defer cancel()
```

benhoyt replied:
Ah, this is a good call -- simplifies things a bit and avoids some extra locking.

hpidcock commented on internal/overlord/checkstate/manager.go:
```diff
@@ -0,0 +1,216 @@
+		switch e := c.lastErr.(type) {
+		case *outputError:
+			info.ErrorDetails = e.output()
+		}
```
Other errors just ignored?

benhoyt replied:
No, other errors are stored in `info.LastError` (the `err.Error()` string). It's just errors that have additional detail that we'll include here -- only the exec check's `outputError` for now.

benhoyt committed:
Service auto-restart and backoffs (refactor as state machine) (#79)

This PR implements auto-restart and backoff functionality for Pebble services.
To achieve this, I've refactored the service handling into a state machine to
make it consistent and easy to add the new states like `backoff` -- most of
`internal/overlord/servstate/handlers.go` has been rewritten.

The PR adds the following new fields to the layer configuration under
services > name (documented in README.md and defaults shown below):

```
on-success: restart
on-failure: restart
backoff-delay: 500ms
backoff-factor: 2.0
backoff-limit: 30s
```

The PR also refactors how/when we track "active services", and means the logs
are saved across service stop/starts (and backoffs/restarts). So we're fixing
a few long-standing issues at once with the refactor.

The list of states is:

```
const (
    stateInitial serviceState = iota
    stateStarting
    stateRunning
    stateTerminating
    stateKilling
    stateStopped
    stateBackoff
)
```

When you start a service it starts at `initial`, then moves to `starting` for
the 1s okay-wait period, then transitions to `running`. When you stop a
service it sends SIGTERM then goes to `terminating` and then `stopped`
(via `killing` if SIGKILL is necessary). 

If a service is configured for auto-restart (the new default), it goes through
the `backoff` state, and there's a `backoffTime` to keep track of the current
backoff time. The new `Backoff*` fields in the service configuration specify
how the backoff delay will advance.

This PR also adds a new status for the `Current` field in the API/CLI:
`backoff`. Example CLI output:

```
$ pebble services
Service  Startup  Current
svc1     enabled  inactive
svc2     enabled  active
svc3     enabled  backoff
```

The PR also adds a `ServiceManager.SendSignal` function to send a signal to a
running service. This is only used for tests right now, but we're planning to
expose it via the API in another PR -- that will be easy to add.

The tests of restart and backoff currently use short timeouts, with some
`time.Sleep()` calls. This seems to work very reliably in my tests, but if it
becomes a problem we change change those to a wait/poll loop to wait for the
logs the service writes, and check the timestamp deltas are what we expect
(but it's a bit more work).

benhoyt committed:
Properly restart service (without backoff) on-check-failure

benhoyt committed:
Merge branch 'master' into custom-health-checks

benhoyt committed:
Add test of long/truncated output in command check

benhoyt committed:
Use strutil.ListContains instead of duplication; go fmt

benhoyt committed:
Make the check have its own context to simplify cancellation

Also add test that cancelling a check in-flight doesn't cause failure
or running the failure action.

benhoyt committed:
Couple of fixes after the merge from master

benhoyt commented:
This is in place
